### PR TITLE
stdlib: add auto formatting to the shell

### PR DESCRIPTION
**Pull request by frazze-jobb**
Adds an option for users to provide a formatter for submitted shell statements.

### CODE REVIEW

**Comment by frazze-jobb** [comment]
Will add an additional case to allow a {Module,Function} tuple that takes a string, in case you want to provide an erlang based formatter

File: lib/kernel/test/interactive_shell_SUITE.erl [hunk 1]
```diff
@@ -477,6 +478,20 @@ shell_multiline_navigation(Config) ->
         stop_tty(Term)
     end.
 
+format_func(String) ->
+    lists:flatten(string:replace(String, "  ", " ", all)).
+
+shell_format(Config) ->
```
Comment (garazdawi) [inline lib/kernel/test/interactive_shell_SUITE.erl:484]: Add a testcase when an external tool is used to format?

File: lib/stdlib/doc/src/stdlib_app.xml [hunk 2]
```diff
@@ -71,6 +71,19 @@
...
+      <item>
+        <p>Can be used to set the formatting of the Erlang shell output. This have an effect
+        on commands that have been submitted, and how it is saved in history. Or if the
+        formatting hotkey is pressed while editing an expression (Alt-f by default). You can specify a Mod:Func/1 that
+        expects the whole expression as a string and returns a formatted expressions as a string.</p>
+        <p>If instead a string is provided, it will be used as a shell command.
+        Your command must include <c>${file}</c>
+        somewhere in the string, for the shell to know where the file goes in the command.
+        Example:
+        -stdlib shell_formatting "\"emacs -batch \${file} -l ~/erlang-format/emacs-format-file -f emacs-format-function\""
```
Comment (garazdawi) [inline lib/stdlib/doc/src/stdlib_app.xml:84]:
```suggestion
        <code>erl -stdlib shell_formatting "\"emacs -batch \${file} -l ~/erlang-format/emacs-format-file -f emacs-format-function\""</code>
```

File: lib/stdlib/doc/src/stdlib_app.xml [hunk 3]
```diff
@@ -71,6 +71,19 @@
       <item>
         <p>Can be used to override the default keymap configuration for the shell.</p>
       </item>
+      <tag><marker id="shell_format_command"/><c>shell_format_command = {Mod, Func} | string() | default</c></tag>
+      <item>
```
Comment (garazdawi) [inline lib/stdlib/doc/src/stdlib_app.xml:75]: Add a cross link to `shell:shell_format_func/1` somewhere in these docs.

And also, should this config variable not be named the same as the function?

File: lib/stdlib/doc/src/shell.xml [hunk 4]
```diff
@@ -993,6 +993,23 @@ q                 - quit erlang
       </desc>
     </func>
 
+    <func>
+      <name name="format_shell_func" arity="1" since="OTP @OTP-18848@"/>
```
Comment (garazdawi) [inline lib/stdlib/doc/src/shell.xml:997]: Add a cross-link to `stdlib_app#format_shell_command` somewhere in these docs.

File: lib/stdlib/doc/src/shell.xml [hunk 5]
```diff
@@ -993,6 +993,23 @@ q                 - quit erlang
...
+      <desc>
+        <p>Can be used to set the formatting of the Erlang shell output. This have an effect
+        on commands that have been submitted, and how it is saved in history. Or if the
+        formatting hotkey is pressed while editing an expression (Alt-f by default). You can specify a Mod:Func/1 that
+        expects the whole expression as a string and returns a formatted expressions as a string.</p>
+        <p>If instead a string is provided, it will be used as a shell command.
+        Your command must include <c>${file}</c>
+        somewhere in the string, for the shell to know where the file goes in the command.
+        Example:
+        shell:format_shell_func("\"emacs -batch \${file} -l ~/erlang-format/emacs-format-file -f emacs-format-function\"").
```
Comment (garazdawi) [inline lib/stdlib/doc/src/shell.xml:1008]:
```suggestion
        <code>shell:format_shell_func("\"emacs -batch \${file} -l ~/erlang-format/emacs-format-file -f emacs-format-function\"").</code>
```

**Review by garazdawi — CHANGES_REQUESTED** [review]
Looks good. Polish tests and docs a bit and this is good to go.

File: lib/stdlib/doc/src/stdlib_app.xml [hunk 6]
```diff
@@ -71,6 +71,20 @@
       <item>
         <p>Can be used to override the default keymap configuration for the shell.</p>
       </item>
+      <tag><marker id="format_shell_func"/><c>format_shell_func = {Mod, Func} | string() | default</c></tag>
+      <item>
+        <p>Can be used to set the formatting of the Erlang shell output. This have an effect
```
Comment (garazdawi) [inline lib/stdlib/doc/src/stdlib_app.xml:76]:
```suggestion
        <p>Can be used to set the formatting of the Erlang shell output. This has an effect
```

File: lib/stdlib/doc/src/stdlib_app.xml [hunk 7]
```diff
@@ -71,6 +71,20 @@
...
+        on commands that have been submitted, and how it is saved in history. Or if the
```
Comment (garazdawi) [inline lib/stdlib/doc/src/stdlib_app.xml:77]:
```suggestion
        on commands that have been submitted and how it is saved in history or if the
```

File: lib/stdlib/doc/src/stdlib_app.xml [hunk 8]
```diff
@@ -71,6 +71,20 @@
...
+        formatting hotkey is pressed while editing an expression (Alt-f by default). You can specify a Mod:Func/1 that
+        expects the whole expression as a string and returns a formatted expressions as a string.
+        See <seemfa marker="shell#format_shell_func/1"><c>shell:format_shell_func/1</c></seemfa>
+        for how to set it from inside the shell.</p>
+        <p>If instead a string is provided, it will be used as a shell command.
+        Your command must include <c>${file}</c>
+        somewhere in the string, for the shell to know where the file goes in the command.
+        </p>
```
Comment (garazdawi) [inline lib/stdlib/doc/src/stdlib_app.xml:85]: Can `erl_pp` be used? If so, maybe add an example?

File: lib/kernel/src/group.erl [hunk 9]
```diff
@@ -886,6 +894,41 @@ get_chars_echo_off1(Drv, Shell) ->
...
+    FormatingCommand = application:get_env(stdlib, format_shell_func),
+    Buffer = edlin:current_line(Cont),
+    try
+        case FormatingCommand of
+            {ok, {M,F}} when is_atom(M), is_atom(F) ->
+                    M:F(Buffer);
+            {ok, FormatingCommand1} when is_list(FormatingCommand1) ->
+                format_expression1(Buffer, FormatingCommand1)
+        end
+    catch
```
Comment (garazdawi) [inline lib/kernel/src/group.erl:907]: If the formatting function crashes, we should log and possibly remove the current formatter?

**Pull request merged**